File: packages/capacitor-actito-inbox/android/src/main/java/com/actito/inbox/capacitor/ActitoInboxPlugin.kt
package com.actito.inbox.capacitor

import android.os.Handler
import android.os.Looper
import androidx.lifecycle.Observer
import com.getcapacitor.*
import com.getcapacitor.annotation.CapacitorPlugin
import com.actito.Actito
import com.actito.ActitoCallback
import com.actito.inbox.ktx.inbox
import com.actito.inbox.models.ActitoInboxItem
import com.actito.models.ActitoNotification
import java.util.*

@CapacitorPlugin(name = "ActitoInboxPlugin")
public class ActitoInboxPlugin : Plugin() {
    private val itemsObserver = Observer<SortedSet<ActitoInboxItem>> { items ->
        if (items == null) return@Observer

        try {
            val data = JSObject()
            data.put("items", JSArray(items.map { it.toJson() }))

            EventBroker.dispatchEvent("inbox_updated", data)
        } catch (e: Exception) {
            logger.error("Failed to emit the inbox_updated event.", e)
        }
    }

    private val badgeObserver = Observer<Int> { badge ->
        if (badge == null) return@Observer

        EventBroker.dispatchEvent("badge_updated", JSObject().apply {
            put("badge", badge)
        })
    }

    override fun load() {
        logger.hasDebugLoggingEnabled = Actito.options?.debugLoggingEnabled ?: false

        EventBroker.setup(this::notifyListeners)

        onMainThread {
            Actito.inbox().observableItems.removeObserver(itemsObserver)
            Actito.inbox().observableItems.observeForever(itemsObserver)

            Actito.inbox().observableBadge.removeObserver(badgeObserver)
            Actito.inbox().observableBadge.observeForever(badgeObserver)
        }
    }

    // region Actito Inbox

    @PluginMethod
    public fun getItems(call: PluginCall) {
        try {
            call.resolve(
                JSObject().apply {
                    put("result", JSArray(Actito.inbox().items.map { it.toJson() }))
                }
            )
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
        }
    }

    @PluginMethod
    public fun getBadge(call: PluginCall) {
        call.resolve(
            JSObject().apply {
                put("result", Actito.inbox().badge)
            }
        )
    }

    @PluginMethod
    public fun refresh(call: PluginCall) {
        Actito.inbox().refresh()
        call.resolve()
    }

    @PluginMethod
    public fun open(call: PluginCall) {
        val json = call.getObject("item") ?: run {
            call.reject("Missing 'item' parameter.")
            return
        }

        val item: ActitoInboxItem = try {
            ActitoInboxItem.fromJson(json)
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
            return
        }

        Actito.inbox().open(item, object : ActitoCallback<ActitoNotification> {
            override fun onSuccess(result: ActitoNotification) {
                try {
                    call.resolve(
                        JSObject().apply {
                            put("result", result.toJson())
                        }
                    )
                } catch (e: Exception) {
                    call.reject(e.localizedMessage)
                }
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun markAsRead(call: PluginCall) {
        val json = call.getObject("item") ?: run {
            call.reject("Missing 'item' parameter.")
            return
        }

        val item: ActitoInboxItem = try {
            ActitoInboxItem.fromJson(json)
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
            return
        }

        Actito.inbox().markAsRead(item, object : ActitoCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun markAllAsRead(call: PluginCall) {
        Actito.inbox().markAllAsRead(object : ActitoCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun remove(call: PluginCall) {
        val json = call.getObject("item") ?: run {
            call.reject("Missing 'item' parameter.")
            return
        }

        val item: ActitoInboxItem = try {
            ActitoInboxItem.fromJson(json)
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
            return
        }

        Actito.inbox().remove(item, object : ActitoCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun clear(call: PluginCall) {
        Actito.inbox().clear(object : ActitoCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    // endregion

    public companion object {
        internal fun onMainThread(action: () -> Unit) = Handler(Looper.getMainLooper()).post(action)
    }
}

File: packages/capacitor-actito-inbox/ios/Sources/ActitoInboxPlugin/ActitoInboxPlugin.swift
import ActitoInboxKit
import ActitoKit
import Capacitor
import Foundation

@objc(ActitoInboxPlugin)
public class ActitoInboxPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "ActitoInboxPlugin"
    public let jsName = "ActitoInboxPlugin"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "getItems", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getBadge", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "refresh", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "open", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "markAsRead", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "markAllAsRead", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "remove", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clear", returnType: CAPPluginReturnPromise),
    ]

    public override func load() {
        addApplicationLaunchListener()

        EventBroker.instance.setup { self.notifyListeners($0, data: $1) }
        Actito.shared.inbox().delegate = self
    }

    @objc func getItems(_ call: CAPPluginCall) {
        do {
            call.resolve([
                "result": try Actito.shared.inbox().items.map { try $0.toJson() }
            ])
        } catch {
            call.reject(error.localizedDescription)
        }
    }

    @objc func getBadge(_ call: CAPPluginCall) {
        call.resolve([
            "result": Actito.shared.inbox().badge
        ])
    }

    @objc func refresh(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            Actito.shared.inbox().refresh { result in
                switch result {
                case .success:
                    call.resolve()
                case let .failure(error):
                    call.reject(error.localizedDescription)
                }
            }
        }
    }

    @objc func open(_ call: CAPPluginCall) {
        guard let json = call.getObject("item") else {
            call.reject("Missing 'item' parameter.")
            return
        }

        let item: ActitoInboxItem

        do {
            item = try ActitoInboxItem.fromJson(json: json)
        } catch {
            call.reject(error.localizedDescription)
            return
        }

        Actito.shared.inbox().open(item) { result in
            switch result {
            case let .success(notification):
                do {
                    call.resolve([
                        "result": try notification.toJson()
                    ])
                } catch {
                    call.reject(error.localizedDescription)
                }
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }

    @objc func markAsRead(_ call: CAPPluginCall) {
        guard let json = call.getObject("item") else {
            call.reject("Missing 'item' parameter.")
            return
        }

        let item: ActitoInboxItem

        do {
            item = try ActitoInboxItem.fromJson(json: json)
        } catch {
            call.reject(error.localizedDescription)
            return
        }

        Actito.shared.inbox().markAsRead(item) { result in
            switch result {
            case .success:
                call.resolve()
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }

    @objc func markAllAsRead(_ call: CAPPluginCall) {
        Actito.shared.inbox().markAllAsRead { result in
            switch result {
            case .success:
                call.resolve()
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }

    @objc func remove(_ call: CAPPluginCall) {
        guard let json = call.getObject("item") else {
            call.reject("Missing 'item' parameter.")
            return
        }

        let item: ActitoInboxItem

        do {
            item = try ActitoInboxItem.fromJson(json: json)
        } catch {
            call.reject(error.localizedDescription)
            return
        }

        Actito.shared.inbox().remove(item) { result in
            switch result {
            case .success:
                call.resolve()
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }

    @objc func clear(_ call: CAPPluginCall) {
        Actito.shared.inbox().clear { result in
            switch result {
            case .success:
                call.resolve()
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }
}

extension ActitoInboxPlugin: ActitoInboxDelegate {
    public func actito(_ actitoInbox: ActitoInbox, didUpdateInbox items: [ActitoInboxItem]) {
        do {
            EventBroker.instance.dispatchEvent("inbox_updated", data: [
                "items": try items.map { try $0.toJson() }
            ])
        } catch {
            logger.error("Failed to emit the inbox_updated event.", error: error)
        }
    }

    public func actito(_ actitoInbox: ActitoInbox, didUpdateBadge badge: Int) {
        EventBroker.instance.dispatchEvent("badge_updated", data: [
            "badge": badge
        ])
    }
}

extension ActitoInboxPlugin {
    private func addApplicationLaunchListener() {
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(didFinishLaunching),
            name: UIApplication.didFinishLaunchingNotification,
            object: nil
        )
    }

    private func removeApplicationLaunchListener() {
        NotificationCenter.default.removeObserver(
            self,
            name: UIApplication.didFinishLaunchingNotification,
            object: nil
        )
    }

    @objc private func didFinishLaunching() {
        removeApplicationLaunchListener()

        logger.hasDebugLoggingEnabled = Actito.shared.options?.debugLoggingEnabled ?? false
    }
}
